### PR TITLE
Tempo Fourier util

### PR DESCRIPTION
**Pull request by Spratiher9**
1-D Fourier transformation capability for Tempo TSDF objects.

Changelog:

* In `python/tempo/tsdf.py` added `function fourier_transform`
* In `python/tempo/utils.py` added new imports
   *  `scipy` and `numpy`
   *  from `scipy` imported `fft` and `fftfreq`
* In `python/tempo/utils.py` added new functions `tempo_fourier_util`, `get_timestep`, `set_timestep`
* In `python/tempo/utils.py` added new constant `TIMESTEP`

For closing issue #96

### CODE REVIEW

**Comment by Spratiher9** [comment]
@rportilla-databricks 

Hi Ricardo, please review the PR.
Thanks,
Souvik

**Review by rportilla-databricks — COMMENTED** [review]
@Spratiher9 , I will review this week - can you please add a unit test for this?

**Comment by Spratiher9** [comment]
Sure @rportilla-databricks I will add unit test and also docstrings for the code.

**Comment by Spratiher9** [comment]
@rportilla-databricks 

Hi Ricardo,
I have added the Fourier transform function's docstrings and the unit test of the same. Please review and let me know in case any changes are required from my side.

**Comment by rportilla-databricks** [comment]
@Spratiher9 could you update the buildtestDF in your test? There are errors in the test which prevent tests from completing: 

1) Update to: 

        # construct dataframes
        df = self.buildTestDF(schema, data, ts_cols=['time'])
        dfExpected = self.buildTestDF(expectedSchema, expected_data, ts_cols=['time'])

2) please add new dependencies in the setup.py as well as requirements (this now requires pyarrow and scipy i believe)

**Comment by Spratiher9** [comment]
@rportilla-databricks 
Ricardo, I have added the changes as you suggested. Please review now.

**Comment by rportilla-databricks** [comment]
Hi @Spratiher9 still getting test failures. Can you share your successful with me? Also, we just added this contribution clause so you can supply information there before I can merge. This was just added, https://github.com/databrickslabs/tempo/blob/master/CONTRIBUTING.md

**Comment by Spratiher9** [comment]
Sure @rportilla-databricks Let me check from my side the tests and check once and also I am adding the updated contributions.md with my information. Meanwhile, Ricardo, Can you share with me the error that you are getting once if you don't mind it will help me in debugging.

**Comment by rportilla-databricks** [comment]
ERROR: test_fourier_transform (__main__.FourierTransformTest)
Test of fourier transform functionality in TSDF objects
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ricardo.portilla/Downloads/mercedes2/tempo/tempo/python/tests/tests.py", line 359, in test_fourier_transform
    self.assertDataFramesEqual(result_tsdf, dfExpected)
  File "/Users/ricardo.portilla/Downloads/mercedes2/tempo/tempo/python/tests/tests.py", line 93, in assertDataFramesEqual
    self.assertSchemasEqual(dfA.schema, dfB.schema)
AttributeError: 'TSDF' object has no attribute 'schema'

----------------------------------------------------------------------
Ran 10 tests in 53.146s

FAILED (errors=1)

**Comment by Spratiher9** [comment]
Hi @rportilla-databricks 
The test is now fixed and the contributions.md also I have added my information.
please review and let me know in case anything else is required.

**Comment by Spratiher9** [comment]
@rportilla-databricks 
Hi Ricardo, 

Now the tests are fixed. The contribution information I have mailed to you. Please go ahead and now review the code.
Let me know in case anything else is required.
Thanks,
Souvik

**Comment by Spratiher9** [comment]
@rportilla-databricks 
Hi Ricardo, Please review the PR now it is ready for release as far as I checked. Please let me know if there is anything missing.

File: python/tempo/tsdf.py [hunk 1]
```diff
@@ -1,533 +1,623 @@
...
+            # timestep = TIMESTEP
+            xf = fftfreq(N[0], timestep)
+            pdf['freq'] = xf
+            return pdf[select_cols + ['freq', 'ft_real', 'ft_imag']]
+
+        valueCol = self.__validated_column(self.df, valueCol)
+        # self.set_timestep(timestep)
+        if self.sequence_col:
+            data = self.df.orderBy(self.ts_col, self.sequence_col)
+            if self.partitionCols == []:
```
Comment (rportilla-databricks) [inline python/tempo/tsdf.py:576]: @Spratiher9 , is this orderBy necessary? Any orderBy or sort is going to incur a big penalty. Instead please sort inside the pandas UDF

File: python/tempo/tsdf.py [hunk 2]
```diff
@@ -1,533 +1,623 @@
...
+                    "tdval", f.col(valueCol))
+                return_schema = ",".join(
+                    [f"{i[0]} {i[1]}" for i in data.dtypes]
+                    +
+                    ["freq double", "ft_real double", "ft_imag double"]
+                )
+                result = data.groupBy(*group_cols).applyInPandas(tempo_fourier_util, return_schema)
+                result = result.drop("tdval")
+        else:
+            data = self.df.orderBy(self.ts_col)
```
Comment (rportilla-databricks) [inline python/tempo/tsdf.py:599]: Same comment here about orderBy.

**Review by rportilla-databricks — CHANGES_REQUESTED** [review]
Please remove orderBy and then we can merge.

File: python/tempo/tsdf.py (hunk 2 above)
Comment (Spratiher9) [reply python/tempo/tsdf.py]: Same here  as before removed orderBy

**Comment by Spratiher9** [comment]
@rportilla-databricks 

Hi Ricardo,
I have added the sort_values inside the pandas udf function and now the values are coming as expected. Please go ahead and review.
Thanks,
Souvik

**Comment by Spratiher9** [comment]
Closing this PR and opening a PR to the release branch instead as per our conversation @rportilla-databricks

**Pull request closed without merging**